stacks-core/src/address.rs

```diff
@@ -0,0 +1,253 @@
+use std::fmt;
+
+use crate::{
+    c32::{decode_address, encode_address},
+    crypto::{
+        hash::{Hash160, SHA256Hash},
+        PublicKey,
+    },
+    StacksError, StacksResult,
+};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AddressHashMode {
+    SerializeP2PKH,
+    SerializeP2SH,
+    SerializeP2WPKH,
+    SerializeP2WSH,
+}
+
+#[derive(Debug, Clone)]
+pub struct StacksAddress {
+    version: u8,
+    hash: Hash160,
+}
+
+impl StacksAddress {
+    pub fn new(version: u8, hash: Hash160) -> Self {
+        Self { version, hash }
+    }
+
+    pub fn version(&self) -> u8 {
+        self.version
+    }
+
+    pub fn hash(&self) -> &Hash160 {
+        &self.hash
+    }
+
+    pub fn from_public_keys(
+        version: u8,
+        public_keys: &[PublicKey],
+        signatures: usize,
+        hash_mode: AddressHashMode,
+    ) -> StacksResult<Self> {
+        let public_key_count = public_keys.len();
+
+        if public_key_count < signatures {
+            return Err(StacksError::InvalidArguments(
+                "Cannot require more signatures than public keys",
+            ));
+        }
+
+        if matches!(
+            hash_mode,
+            AddressHashMode::SerializeP2PKH | AddressHashMode::SerializeP2WPKH
+        ) {
+            if public_key_count != 1 {
+                return Err(StacksError::InvalidArguments(
+                    "Cannot use more than one public key for this hash mode",
+                ));
+            }
+
+            if signatures != 1 {
+                return Err(StacksError::InvalidArguments(
+                    "Cannot require more than one signature for this hash mode",
+                ));
+            }
+        }
+
+        let hash = match hash_mode {
+            AddressHashMode::SerializeP2PKH => hash_p2pkh(&public_keys[0]),
+            AddressHashMode::SerializeP2SH => hash_p2sh(signatures, public_keys),
+            AddressHashMode::SerializeP2WPKH => hash_p2wpkh(&public_keys[0]),
+            AddressHashMode::SerializeP2WSH => hash_p2wsh(signatures, public_keys),
+        };
+
+        Ok(Self::new(version, hash))
+    }
+}
+
+impl TryFrom<&StacksAddress> for String {
+    type Error = StacksError;
+
+    fn try_from(address: &StacksAddress) -> Result<Self, Self::Error> {
+        encode_address(address.version, address.hash.as_ref()).map_err(|err| err.into())
+    }
+}
+
+impl TryFrom<&str> for StacksAddress {
+    type Error = StacksError;
+
+    fn try_from(address: &str) -> Result<Self, Self::Error> {
+        let (version, hash) =
+            decode_address(address).map_err::<StacksError, _>(|err| err.into())?;
+
+        Ok(Self::new(version, Hash160::new(hash)))
+    }
+}
```

Comment:
Are these traits necessary? Otherwise I'd prefer to rely on the Display and FromStr traits which would be more idiomatic for string representation and parsing.

Reply:
Both conversions are fallible, and that's hidden in the `blockstack-lib` implementation with unwraps. I believe this to be closer to the truth. It also prevents the user to be confused, compared to panicking.